openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	workloadAnnotations := getWorkloadAnnotations(pod.Annotations)
+
+	// strip any resource annotations specified by a user
+	stripResourcesAnnotations(pod.Annotations)
+	// strip any workload annotation to prevent from underlying components(CRI-O, kubelet) to apply any changes
+	// according to the workload annotation
+	stripWorkloadAnnotations(pod.Annotations)
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotations) > 1 {
+		pod.Annotations[workloadAdmissionError] = fmt.Sprintf(admissionErrorMultipleWorkloadAnnotations, workloadAnnotations)
+		return nil
+	}
+
+	workloadType, err := getWorkloadType(workloadAnnotations)
+	if err != nil {
+		pod.Annotations[workloadAdmissionError] = fmt.Sprintf(admissionErrorFailedGetWorkloadType, err)
+		return nil
+	}
+
+	// no workload annotation is available
+	if workloadType == "" {
+		return nil
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	// TODO: currently we do not check the effect, but it can be useful in the future
+	_, err = getWorkloadAnnotationEffect(workloadAnnotations[workloadAnnotation])
+	if err != nil {
+		pod.Annotations[workloadAdmissionError] = fmt.Sprintf(admissionErrorFailedGetWorkloadAnnotationEffect, err)
+		return nil
+	}
+
+	if !a.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(attr, fmt.Errorf("%s node and namespace caches not synchronized", PluginName))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace %q: %v", PluginName, attr.GetNamespace(), err)
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if !isNamespaceAllowed(ns.Annotations, workloadType) {
+		return nil
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if len(nodes) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("%s the cluster does not have any nodes", PluginName))
+	}
+
+	// TODO: currently we supports only SNO use case because we have not yet worked out the best approach to determining whether the feature
+	// should be on or off in a multi-node cluster, and computing that state incorrectly could lead to breaking running clusters.
+	if len(nodes) > 1 {
+		pod.Annotations[workloadAdmissionError] = admissionErrorOnlySNOSupported
```

Comment:
This creates a useless annotation on each pod in normal non-single-node mode?

Reply:
in general yes, so I will remove the annotation in such use case

Reply:
done